primo/opt_model/result_parser.py

```diff
@@ -0,0 +1,804 @@
+
+LOGGER = logging.getLogger(__name__)
+
+
+class OptimalProject:
+    """
+    Class for storing optimal projects
+    """
+
+    def __init__(
+        self, wd: WellData, index: list, plugging_cost: float, project_id: int
+    ):
+        """
+        Constructs an object for storing optimal project results
+        Parameters
+        ----------
+        wd : WellData
+            A WellData object
+
+        index : list
+            List of indices/rows/wells belonging to the cluster/group
+
+        plugging_cost : float
+            Cost of plugging
+
+        project_id : int
+            Project id
+        """
+        # Motivation for storing the entire DataFrame: After the problem is solved
+        # it is desired to display/highlight flagged wells for which the information
+        # was not available, and the missing data was filled. Having the entire
+        # DataFrame allows ease of access to flagged wells (columns containing _flag)
+        self.well_data = wd._construct_sub_data(index)
+        self._col_names = self.well_data._col_names
+        col_names = self._col_names
+        self.project_id = project_id
+
+        # Must display essential columns while printing a DataFrame
+        self._essential_cols = [
+            col_names.well_id,
+            col_names.operator_name,
+            col_names.latitude,
+            col_names.longitude,
+            col_names.age,
+            col_names.depth,
+            col_names.priority_score,
+        ]
+        self._priority_score_cols = wd.get_priority_score_columns
+        self._flag_cols = wd.get_flag_columns
+        self.num_wells = len(index)
+        # Optimization problem uses million USD. Convert it to USD
+        self.plugging_cost = plugging_cost * 1e6
+        self._add_distance_to_centroid_col()
+        self.efficiency_score = 0
+        accessibility_column_attr = ["elevation_delta", "dist_to_road"]
+        self.accessibility_attr = [
+            attribute
+            for attribute in accessibility_column_attr
+            if hasattr(col_names, attribute)
+        ]
+
+    def __iter__(self):
+        return iter(self.well_data.data.index)
+
+    def __str__(self) -> str:
+        msg = (
+            f"Number of wells in project {self.project_id}\t\t: {self.num_wells}\n"
+            f"Estimated Project Cost\t\t\t: ${round(self.plugging_cost)}\n"
+            f"Impact Score [0-100]\t\t\t: {self.impact_score}\n"
+            f"Efficiency Score [0-100]\t\t: {self.efficiency_score}\n"
+        )
+        return msg
+
+    def _check_column_exists(self, col_name):
+        """
+        Checks if a column exists
+        """
+        if col_name is None:
+            raise ValueError(f"Information for column {col_name} is not available.")
+
+    @property
+    def num_wells_near_hospitals(self):
+        """Returns number of wells that are near hospitals"""
+        col_name = self._col_names.hospitals
+        self._check_column_exists(col_name)

```

Comment:
One note here: this would only apply if a person was trying to access the attribute "num_wells_near_hospitals", which if they don't have the data means that they are doing something unintentional. I think I am of the opinion that it should be an error, because it means either the user has the data and forgot to add it in (in which case throwing an error is correct) or the user has somehow used this attribute without actually having data (for which a print statement may suffice, but really the user's code calling the attribute is wrong). We can do internal checks to make sure that the data exists when the attribute is called.

Reply:
I understand the motivation now. I'm okay with raising an error. What we need to decide is whether it is going to be `pandas` `KeyError` or a custom error message. Certainly a custom error message would be helpful. But then it needs to be more informative than pandas KeyError.